### ISSUE TITLE
LogContext throws SecurityException when making cross AppDomain calls in AppDomain with limited permissions

### ISSUE THREAD

**Issue by zacharycmontoya** [serilog/serilog#1486]
**Description**
On .NET Framework 4.5.x, the `LogContext` is stored in the `CallContext` because there is not an AsyncLocal primitive. As a result of storing the object in the `CallContext` and invoking the .NET Remoting infrastructure, issue #987 began occurring with a fix being issued in #992 that was published in Serilog 2.6.0. That original problem was resolved, but it created a new problem: the solution to keeping Serilog objects in the `CallContext` alive so they are not disconnected by the "server" introduced new logic into the object that was transmitted to the server so that it would keep itself alive. However, in the case of a cross AppDomain call where the new AppDomain does not have permission to use the remoting infrastructure (`SecurityPermissionFlag.RemotingConfiguration`), this lifetime management fails when the permission is demanded, with the following error message:

```
Unhandled Exception: System.Security.SecurityException: Request for the permission of type 'System.Security.Permissions.SecurityPermission, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' failed.
```

**Reproduction**
Please provide code samples showing how you're configuring and calling Serilog to produce the behavior.
```csharp
        public static void Main(string[] args)
        {
            // Set up an AppDomain with only code execution permissions
            // Notably, it lacks the SecurityPermissionFlag.RemotingConfiguration flag
            var permSet = new PermissionSet(PermissionState.None);
            permSet.AddPermission(new SecurityPermission(SecurityPermissionFlag.Execution));
            var remote = AppDomain.CreateDomain("Remote", null, AppDomain.CurrentDomain.SetupInformation, permSet);

            try
            {
                using (LogContext.PushProperty("Prop1", 42))
                {
                    remote.DoCallBack(HelloWorld);
                }
            }
            finally
            {
                AppDomain.Unload(remote);
            }
        }

        public static void HelloWorld()
        {
            Console.WriteLine("Hello World");
        }
```

**Expected behavior**
The program should execute without any issues, and the cross-AppDomain call should write "Hello World" to the console.

**Relevant package, tooling and runtime versions**
- Serilog versions 2.6.0-2.10.0
- .NET Framework 4.5.x (which use the `CallContext` to store ambient `LogContext` information)

**Additional context**
This problem was encountered in the DataDog/dd-trace-dotnet repo in https://github.com/DataDog/dd-trace-dotnet/pull/925, so I will follow up with a proposed solution and see if it is appropriate for this library. The solution we implemented there was to modify the `DisposableObjectHandle` that is stored in the `CallContext` so that its override of `InitializeLifetimeService` returns `null` instead of invoking any remoting infrastructure. This avoids the remoting permission issue at the cost of keeping the object alive indefinitely. However, this is easily handled by disconnecting the object ourselves when `Dispose()` is called.

**Comment by nblumhardt** [comment]
Hi Zach! Thanks for the report and analysis. If you're able to figure out a solution that doesn't regress other existing usages of the library, it would be welcome 👍

**Comment by nblumhardt** [comment]
Hi Zach - any updates on how the fix went in dd-trace? (No other regressions noticed since then?) :-)

**Comment by zacharycmontoya** [comment]
Hi @nblumhardt , thanks for the bump 😄 The fix in `dd-trace-dotnet` has been good and we haven't had any reported issues since, so I've ported the fix in #1584. Please let me know if you have any feedback

**Comment by nblumhardt** [comment]
Discussion on the linked PR concluded with `wontfix`
